{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Webscraping for the Cincinnati Children's Hospital Medical Center LOVD \n",
    "The CCHMC LOVD page is still on the LOVD version 2 format, so it is a little bit more difficult to scrape than the LOVD3 sites. It is possible to see some information about the individual variants, but the clinical information is all obtained on a separate page for each individual variant. That said, I have to scrape one web page to find the subsequent pages, and then I have to request those pages and scrape the information from each of those. Additionally, the number of fields on those subsequent pages can vary, so I ended up putting the data into a dictionary and then putting a dash for the fields that were missing. Once the data from webscraping is obtained, I have to read in a file that gives the versions of transcripts that Biocommons has (because it will throw errors with the ones the CCHMC site gives) and then converts the cDNA information to genomic information. After that, the genomic information is used to populate fields that can be obtained in ClinVar and the output is put in a format that mirrors the ClinVar output as close as I was able to make it.   \n",
    "\n",
    "The base page is [this](https://research.cchmc.org/LOVD2/variants.php?select_db=ACADM&action=search_all&order=Variant%2FDNA%2CASC&hide_col=&show_col=&limit=1000&search_pathogenic_=&search_Variant%2FDNA=&search_Variant%2FProtein=&search_Variant%2FRemarks=&search_Variant%2FdbSNP=&search_Patient%2FPatient%2FRace=&search_Patient%2FReference=&search_Patient%2FOrigin%2FEthnic=). \n",
    "\n",
    "You can find the transcript, the last updated date and the chromosome on this [page](https://research.cchmc.org/LOVD2/home.php?select_db=ACADM). \n",
    " \n",
    "\n",
    "For now, the first cell lists the genes of interest. This will probably be converted to a .py file and we will move the list of genes out to the command line so that you can include any genes of interest, but for now, it is the first cell. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_for_metabolic_diseases = [\"ACADM\", \"ACADS\", \"ACADVL\", \"CPT1A\", \"CPT2\", \"ETFA\", \"ETFB\", \"ETFDH\", \"HADH\",\n",
    "                               \"HADHA\", \"HADHB\", \"SLC22A5\", \"SLC25A20\"]\n",
    "SCID_genes = [\"CORO1A\", \"CD3Z\", \"ATM\", \"LIG4\", \"DOCK2\", \"PRKDC\", \"DCLRE1C\", \"CHD7\", \"NHEJ1\", \"GATA2\",\n",
    "            \"TBX1\", \"IL2RG\", \"MTHFD1\", \"RAC2\", \"IGHM\", \"ADA\", \"IL7RA\", \"MTR\", \"CD3G\", \"BTK\",\n",
    "            \"AK2\", \"JAK3\", \"RMRP\", \"STAT5B\", \"CD40LG\", \"CD3D\", \"RAG1\", \"SLC46A1\", \"ZAP70\", \"WAS\",\n",
    "            \"CD3E\", \"RAG2\", \"DOCK8\", \"PNP\", \"DKC1\", \"PTPRC\", \"FOXN1\", \"NBN\", \"BLNK\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "ncbi_api_key = os.environ.get('ncbi_api_key') # I don't think this is doing anything\n",
    "# this should theoretically speed it up to 10 per second instead of 3 per second\n",
    "# https://github.com/biocommons/bioutils\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "import warnings \n",
    "import requests \n",
    "import json \n",
    "import hgvs.normalizer\n",
    "import hgvs.variantmapper\n",
    "import hgvs.dataproviders.uta\n",
    "import hgvs.assemblymapper\n",
    "import hgvs.parser\n",
    "import hgvs.validator\n",
    "import hgvs.exceptions\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "hp = hgvs.parser.Parser()\n",
    "hn = hgvs.normalizer.Normalizer(hdp)\n",
    "am = hgvs.assemblymapper.AssemblyMapper(hdp, assembly_name='GRCh37', alt_aln_method='splign', replace_reference=True)\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "from datetime import datetime\n",
    "\n",
    "output_directory = \"All_ClinVar_output_files/\"\n",
    "os.makedirs(output_directory+\"ClinVar/Not_Valid_Annotations/\", exist_ok=True)\n",
    "\n",
    "from datetime import datetime\n",
    "\n",
    "ncbi_api_key = os.environ.get('ncbi_api_key')\n",
    "output_directory = \"output_files/\"\n",
    "os.makedirs(output_directory+\"CCHMC_LOVD/\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts_df = pd.read_csv(\"Transcript_Info_For_Dictionaries.csv\", index_col = 0) #, header = None, skiprows=1, index_col = 0\n",
    "transcript_dictionary = dict(zip(list(transcripts_df[\"Transcript\"].values), list(transcripts_df[\"Transcript_ID_Full\"].values)))\n",
    "gene_to_transcript_dict = dict(zip(list(transcripts_df[\"Gene\"].values), list(transcripts_df[\"Transcript_ID_Full\"].values)))\n",
    "\n",
    "def web_scrape_cchmc(disease, list_of_genes):\n",
    "    for gene in list_of_genes:\n",
    "        result_list = []\n",
    "        home_link = \"https://research.cchmc.org/LOVD2/home.php?select_db=\"+gene\n",
    "        response = requests.get(home_link)\n",
    "        response_html = response.content.decode(\"utf-8\")\n",
    "        bs_format = BeautifulSoup(response_html, \"html.parser\")\n",
    "        potential_error_message = bs_format.select('th')[0].get_text()\n",
    "        if potential_error_message == 'Error: Init':\n",
    "            print(\"The CCHMC LOVD2 database does not contain any variants for the gene \"+gene+\".\", sep = \"\")\n",
    "            # I could change this out for a warning, but the CCHMC database doesn't contain any info for the majority of the genes\n",
    "            # so I thought this would be more appropriate to standard output instead of a warning to standard error\n",
    "            continue\n",
    "\n",
    "        updated_date = \"-\"\n",
    "        chromosome = \"-\"\n",
    "        transcript = \"-\"\n",
    "        transcript_new = \"-\"\n",
    "        genomic_ref = \"-\" # This is still referring to the gene so I don't think we will use it, but I am grabbing it for just in case\n",
    "        for term in bs_format.select('th'):\n",
    "            term_text = term.get_text()\n",
    "\n",
    "            if term_text == \"Last update\" :\n",
    "                updated_date = term.parent.select('td')[0].get_text()\n",
    "            elif term_text == \"Chromosome Location\":\n",
    "                chromosome_search_result = re.search(r\"([0-9,X,Y,M,T]+)[p|q]\\d+\", term.parent.select('td')[0].get_text())\n",
    "                if chromosome_search_result != None:\n",
    "                    chromosome = chromosome_search_result.group(1)\n",
    "            elif \"Transcript\" in term_text:\n",
    "                transcript = term.parent.select('td')[0].get_text()\n",
    "            elif \"Genomic\" in term_text:\n",
    "                genomic_ref = term.parent.select('td')[0].get_text()\n",
    "        transcript_seach = re.search(r\"NM_\\d+\", transcript)\n",
    "        if transcript_seach != None:\n",
    "            transcript_parent = transcript_seach.group(0)\n",
    "            transcript_new = transcript_dictionary[transcript_parent]\n",
    "        ##\n",
    "        ## The ones that had no transcript listed on the home page only have one transcript\n",
    "        ## so I am going to use the dictionary to look up the transcript for those ones\n",
    "        ## \n",
    "        if transcript_new == \"-\":\n",
    "            transcript_new = gene_to_transcript_dict[gene]\n",
    "\n",
    "        link = \"https://research.cchmc.org/LOVD2/variants.php?select_db=\"+gene+\"&action=search_all&order=Variant%2FDNA%2CASC&hide_col=&show_col=&limit=1000&search_pathogenic_=&search_Variant%2FDNA=&search_Variant%2FProtein=&search_Variant%2FRemarks=&search_Variant%2FdbSNP=&search_Patient%2FPatient%2FRace=&search_Patient%2FReference=&search_Patient%2FOrigin%2FEthnic=\"\n",
    "        response = requests.get(link)\n",
    "        if response.status_code == 200:\n",
    "            response_html = response.content.decode(\"utf-8\")\n",
    "            bs_format = BeautifulSoup(response_html, \"html.parser\")\n",
    "            potential_error_message = bs_format.select('img[src=\"./gfx/header_variant_listings.png\"]')[0].parent.parent.select('td')[0].get_text().strip()\n",
    "            if \"currently no public variants\" in potential_error_message:\n",
    "                print(\"The CCHMC LOVD2 database does not contain any variants for the gene \"+gene+\".\", sep = \"\")\n",
    "                continue\n",
    "                # This should not happen because it should have happened with the first link if this were the case. \n",
    "\n",
    "            links = []\n",
    "            anchor_data = bs_format.select('a[class=\"data\"]')\n",
    "            for anchor in anchor_data:\n",
    "                links.append(\"https://research.cchmc.org\"+anchor.attrs['href'])\n",
    "            result_list = []\n",
    "            \n",
    "            for link in links:\n",
    "                response = requests.get(link)\n",
    "                if response.status_code != 200:\n",
    "                    continue\n",
    "                    # I am not going to write a warning statement for any of these \n",
    "                    # because if they fail we only miss out on a single variant\n",
    "                response_html = response.content.decode(\"utf-8\")\n",
    "                bs_format = BeautifulSoup(response_html, \"html.parser\")\n",
    "                ths = bs_format.select('th[valign=\"top\"]')\n",
    "                individual_results = []\n",
    "                individual_labels = []\n",
    "                for th in ths:\n",
    "                    # This can get all of the column values for the things we care about in a list format\n",
    "                    # Unfortunately, some variables are missing for some variants\n",
    "                    individual_results.append(th.parent.td.get_text().replace(u'\\xa0', u' ').replace('(View in UCSC Genome Browser, Ensembl)', ''))\n",
    "                for label in bs_format.select('th[valign=\"top\"]'):\n",
    "                    # This will be the labels that correspond with the values obtained above. \n",
    "                    individual_labels.append(label.get_text().replace(u'\\xa0', u' '))\n",
    "                # I have the values and their labels in separate lists at this point. \n",
    "                # Now I am going to put them into a dictionary so that I can have key, value pairs\n",
    "                # and I can put in a dash for anything that isn't present\n",
    "                individual_dictionary = dict(zip(individual_labels, individual_results))\n",
    "                if \"Race\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Race\"] = \"-\"\n",
    "                if \"Reference\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Reference\"] = \"-\"\n",
    "                if \"Ethnic origin\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Ethnic origin\"] = \"-\"\n",
    "                if \"Submitter\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Submitter\"] = \"-\"\n",
    "                if \"Allele\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Allele\"] = \"-\"\n",
    "                if \"Reported pathogenicity\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Reported pathogenicity\"] = \"-\"\n",
    "                if \"Concluded pathogenicity\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Concluded pathogenicity\"] = \"-\"\n",
    "                if \"DNA change\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"DNA change\"] = \"-\"\n",
    "                if \"Protein\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Protein\"] = \"-\"\n",
    "                if \"Variant remarks\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"Variant remarks\"] = \"-\"\n",
    "                if \"dbSNP ID\" not in individual_dictionary:\n",
    "                    individual_dictionary[\"dbSNP ID\"] = \"-\"\n",
    "\n",
    "                # Now I am going to convert the pathogenicity to the ClinVar terms\n",
    "                pathogenicity_clinvar = \"-\"\n",
    "                if individual_dictionary[\"Reported pathogenicity\"] == \"No known pathogenicity\":\n",
    "                    pathogenicity_clinvar = \"Benign\"\n",
    "                elif individual_dictionary[\"Reported pathogenicity\"] == \"Probably no pathogenicity\":\n",
    "                    pathogenicity_clinvar = \"Likely benign\"\n",
    "                elif individual_dictionary[\"Reported pathogenicity\"] == \"Unknown\":\n",
    "                    pathogenicity_clinvar = \"Uncertain significance\"\n",
    "                elif individual_dictionary[\"Reported pathogenicity\"] == \"Probably pathogenic\":\n",
    "                    pathogenicity_clinvar = \"Likely pathogenic\"\n",
    "                elif individual_dictionary[\"Reported pathogenicity\"] == \"Pathogenic\":\n",
    "                    pathogenicity_clinvar = \"Pathogenic\"\n",
    "\n",
    "\n",
    "                ###\n",
    "                ###  Now I need to see if I can use biocommons to get the genomic annotation\n",
    "                ###  I don't need to keep the rest of these in a dictionary because I can save them as variables\n",
    "                ###          \n",
    "                hgvs_transcript = \"-\"\n",
    "                Genomic_annotation = \"-\"\n",
    "                variant = individual_dictionary[\"DNA change\"].strip() # This should always be present, even if it is a dash\n",
    "                if transcript_new != \"-\" and variant != \"-\":\n",
    "                    hgvs_transcript = transcript_new+\":\"+variant\n",
    "                elif transcript != \"-\" and variant != \"-\":\n",
    "                    # if there is no new transcript, give it a test with the old one\n",
    "                    hgvs_transcript = transcript+\":\"+variant\n",
    "                    # Leave it as a dash if the DNA change is a dash, which is common\n",
    "                if hgvs_transcript != \"-\":\n",
    "                    try:\n",
    "                        Genomic_annotation = str(am.c_to_g(hp.parse_hgvs_variant(hgvs_transcript)))\n",
    "                    except hgvs.exceptions.HGVSError:\n",
    "                        Genomic_annotation = \"-\"\n",
    "                Position_g_start = \"-\"\n",
    "                Position_g_stop = \"-\"\n",
    "                Ref_allele = \"-\"\n",
    "                Alt_allele = \"-\"\n",
    "                Var_Type = \"-\"\n",
    "                Chr_Accession = \"-\"\n",
    "                Var_Length = \"-\"\n",
    "\n",
    "\n",
    "                if Genomic_annotation != \"-\":\n",
    "                    # If we have the genomic annotation, we can use this to populate the fileds that are empty\n",
    "                     # If we can't get a genomic annotation, then all the other variables will be left as a dash\n",
    "                    chrom_accession_search = re.search(r\"NC_\\d+\\.\\d+\", Genomic_annotation)\n",
    "                    if chrom_accession_search != None:\n",
    "                        Chr_Accession =chrom_accession_search.group(0)\n",
    "                    if \">\" in Genomic_annotation:\n",
    "                        Var_Type = \"single nucleotide variant\"\n",
    "                    elif \"delins\" in Genomic_annotation:\n",
    "                        Var_Type = \"Indel\"\n",
    "                    elif \"del\" in Genomic_annotation:\n",
    "                        Var_Type = \"Deletion\"\n",
    "                    elif \"ins\" in Genomic_annotation:\n",
    "                        Var_Type = \"Insertion\"\n",
    "                    elif \"dup\" in Genomic_annotation:\n",
    "                        Var_Type = \"Duplication\"\n",
    "                    elif \"inv\" in Genomic_annotation:\n",
    "                        Var_Type = \"Inversion\"\n",
    "                    # These have all been normalized by biocommons, so this should work. \n",
    "\n",
    "                    pos_start_search = re.search(r\"g.(\\d+)\", Genomic_annotation)\n",
    "                    if pos_start_search != None:\n",
    "                        Position_g_start = pos_start_search.group(1)\n",
    "                    if Var_Type == \"single nucleotide variant\":\n",
    "                        Position_g_stop = Position_g_start\n",
    "                        Var_Length = 1\n",
    "                        ref_allele_search = re.search(r\"g.\\d+([A,C,T,G])>([A,C,T,G])\", Genomic_annotation)\n",
    "                        if ref_allele_search != None:\n",
    "                            Ref_allele = ref_allele_search.group(1)\n",
    "                            Alt_allele = ref_allele_search.group(2)\n",
    "                    elif Var_Type == \"Indel\":\n",
    "                        # Var_Length for indels is inconsistent in ClinVar, so I will leave these as a dash\n",
    "                        alt_allele_search = re.search(r\"ins([A,C,T,G]+)\", Genomic_annotation)\n",
    "                        if alt_allele_search != None:\n",
    "                            Alt_allele = alt_allele_search.group(1)\n",
    "                        multiple_location_search = re.search(r\"\\d+_\\d+\", Genomic_annotation)\n",
    "                        if multiple_location_search == None:\n",
    "                            Position_g_stop = Position_g_start      \n",
    "                        else:\n",
    "                            stop_search = re.search(r\"_(\\d+)del\", Genomic_annotation)\n",
    "                            if stop_search != None:\n",
    "                                Position_g_stop = stop_search.group(1)\n",
    "                    elif Var_Type == \"Insertion\":\n",
    "                        # Reference will always be a dash\n",
    "                        # I think the stop should always be start plus one, but in case it is funky I am gong to use the regex anyway\n",
    "                        stop_search = re.search(r\"_(\\d+)[a-z]+\", Genomic_annotation)\n",
    "                        if stop_search != None:\n",
    "                            Position_g_stop = stop_search.group(1)\n",
    "                        # Ref allele is always a dash, and we can get the Alt allele from the end of this \n",
    "                        alt_allele_search = re.search(r\"ins([A,C,T,G]+)\", Genomic_annotation)\n",
    "                        if alt_allele_search != None:\n",
    "                            Alt_allele = alt_allele_search.group(1)\n",
    "                        if Alt_allele != \"-\":\n",
    "                            Var_Length = len(Alt_allele)\n",
    "                    elif Var_Type == \"Duplication\" or Var_Type == \"Deletion\" or Var_Type == \"Inversion\":\n",
    "                        # Can't get reference or alternate for either of these\n",
    "                        multiple_location_search = re.search(r\"\\d+_\\d+\", Genomic_annotation)\n",
    "                        if multiple_location_search == None:\n",
    "                            Position_g_stop = Position_g_start \n",
    "                            Var_Length = 1\n",
    "                        else:\n",
    "                            stop_search = re.search(r\"_(\\d+)[a-z]+\", Genomic_annotation)\n",
    "                            if stop_search != None:\n",
    "                                Position_g_stop = stop_search.group(1)\n",
    "                                Var_Length = int(Position_g_stop) - int(Position_g_start) + 1\n",
    "\n",
    "                # Genomic annotation goes in twice for the genomic annotation and HGVS normalized genomic annotation (to match ClinVar output format)\n",
    "                # gene will also be in here twice for Affected Genes and Gene Symbol\n",
    "                individual_variant_info = [\"GRCh37\", chromosome, Position_g_start, Position_g_stop, Ref_allele,\n",
    "                                          Alt_allele, Genomic_annotation, Genomic_annotation, Var_Type, Var_Length,\n",
    "                                          pathogenicity_clinvar, \"-\", individual_dictionary[\"Allele\"],\n",
    "                                          \"-\", gene, gene, \"-\", transcript_new, variant, \n",
    "                                          hgvs_transcript, \"-\", individual_dictionary[\"Protein\"], \"-\", Chr_Accession,\n",
    "                                          \"-\", \"-\", \"-\", \"CCHMC_LOVD2\", \"-\", \"-\", \"-\", individual_dictionary[\"Submitter\"], \n",
    "                                          updated_date, individual_dictionary[\"Concluded pathogenicity\"],\n",
    "                                           individual_dictionary[\"Reported pathogenicity\"], individual_dictionary[\"Race\"],\n",
    "                                          individual_dictionary[\"Ethnic origin\"], individual_dictionary[\"Reference\"], \n",
    "                                          individual_dictionary[\"Variant remarks\"], transcript, individual_dictionary[\"dbSNP ID\"],\n",
    "                                          genomic_ref]\n",
    "                result_list.append(individual_variant_info)\n",
    "        gene_df = pd.DataFrame(result_list)\n",
    "        # The added columns starting with \"CCHMC Reported Pathogenicity\" are all new to this one and not in ClinVar\n",
    "        gene_df.columns = [\"Genome Assembly\", \"Chr\", \"Position Start\", \"Position Stop\", \"Ref\", \"Alt\",\n",
    "                           \"Genomic Annotation\", \"HGVS Normalized Genomic Annotation\", \"Variant Type\",\n",
    "                            \"Variant Length\", \"Pathogenicity\", \"Disease\", \"Genetic Origin\", \"Inheritance Pattern\",\n",
    "                            \"Affected Genes\" , \"Gene Symbol\", \"Compound Het Status\", \"Transcript\",\n",
    "                            \"Transcript Notation\", \"HGVS Transcript Notation\",\"Protein Accession\",\n",
    "                            \"Protein Notation\", \"HGVS Protein Annotation\", \"Chr Accession\", \"VCF Pos\",\n",
    "                            \"VCF Ref\", \"VCF Alt\", \"Database\", \"ClinVar Accession\", \"Review Status\",\n",
    "                            \"Star Level\", \"Submitter\", \"Edited Date\", \"CCHMC Reported Pathogenicity\", \n",
    "                           \"CCHMC Concluded Pathogenicity\", \"Race\", \"Ethnic Origin\", \"Reference Paper\", \n",
    "                          \"Variant Remarks\", \"CCHMC Reported Transcript\", \"dbSNP ID\", \"Genomic Accession\"]\n",
    "        gene_df.to_csv(output_directory+\"CCHMC_LOVD/\"+disease+\"_\"+gene+\"_CCHMC_results.csv\")\n",
    "        print(\"Finished obtaining variant information for gene \"+gene+\" in the CCHMC database.\", sep = \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished obtaining variant information for gene ACADM in the CCHMC database.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ACADS.\n",
      "Finished obtaining variant information for gene ACADVL in the CCHMC database.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CPT1A.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CPT2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ETFA.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ETFB.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ETFDH.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene HADH.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene HADHA.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene HADHB.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene SLC22A5.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene SLC25A20.\n"
     ]
    }
   ],
   "source": [
    "web_scrape_cchmc(\"Metabolic_Diseases\", genes_for_metabolic_diseases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The CCHMC LOVD2 database does not contain any variants for the gene CORO1A.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CD3Z.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ATM.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene LIG4.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene DOCK2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene PRKDC.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene DCLRE1C.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CHD7.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene NHEJ1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene GATA2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene TBX1.\n",
      "Finished obtaining variant information for gene IL2RG in the CCHMC database.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene MTHFD1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene RAC2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene IGHM.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ADA.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene IL7RA.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene MTR.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CD3G.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene BTK.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene AK2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene JAK3.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene RMRP.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene STAT5B.\n",
      "Finished obtaining variant information for gene CD40LG in the CCHMC database.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CD3D.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene RAG1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene SLC46A1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene ZAP70.\n",
      "Finished obtaining variant information for gene WAS in the CCHMC database.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene CD3E.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene RAG2.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene DOCK8.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene PNP.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene DKC1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene PTPRC.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene FOXN1.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene NBN.\n",
      "The CCHMC LOVD2 database does not contain any variants for the gene BLNK.\n"
     ]
    }
   ],
   "source": [
    "web_scrape_cchmc(\"SCID\", SCID_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished obtaining variant information for gene IL2RG in the CCHMC database.\n",
      "Finished obtaining variant information for gene WAS in the CCHMC database.\n"
     ]
    }
   ],
   "source": [
    "failed_genes_list = [\"IL2RG\", \"WAS\"]\n",
    "web_scrape_cchmc(\"SCID\", failed_genes_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
